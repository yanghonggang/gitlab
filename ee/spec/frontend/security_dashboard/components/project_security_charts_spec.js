import { GlLoadingIcon } from '@gitlab/ui';
import { GlLineChart } from '@gitlab/ui/dist/charts';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import VueApollo from 'vue-apollo';
import DashboardNotConfigured from 'ee/security_dashboard/components/empty_states/reports_not_configured.vue';
import ProjectSecurityCharts from 'ee/security_dashboard/components/project_security_charts.vue';
import SecurityChartsLayout from 'ee/security_dashboard/components/security_charts_layout.vue';
import projectsHistoryQuery from 'ee/security_dashboard/graphql/project_vulnerabilities_by_day_and_count.graphql';
import createMockApollo from 'jest/helpers/mock_apollo_helper';
import {
  mockProjectSecurityChartsWithData,
  mockProjectSecurityChartsWithoutData,
} from '../mock_data';

const localVue = createLocalVue();
localVue.use(VueApollo);

describe('Project Security Charts component', () => {
  let wrapper;

  const projectFullPath = 'project/path';
  const helpPath = 'docs/security/dashboard';

  const findLineChart = () => wrapper.find(GlLineChart);
  const findLoadingIcon = () => wrapper.find(GlLoadingIcon);
  const findEmptyState = () => wrapper.find(DashboardNotConfigured);

  const createApolloProvider = (...queries) => {
    return createMockApollo([...queries]);
  };

  const createComponent = ({ query, propsData, chartWidth = 1024 }) => {
    const component = shallowMount(ProjectSecurityCharts, {
      localVue,
      apolloProvider: createApolloProvider([
        projectsHistoryQuery,
        jest.fn().mockResolvedValue(query),
      ]),
      propsData: {
        projectFullPath,
        helpPath,
        ...propsData,
      },
      stubs: {
        SecurityChartsLayout,
      },
    });

    // Need to setData after the component is mounted
    component.setData({ chartWidth });

    return component;
  };

  afterEach(() => {
    if (wrapper) {
      wrapper.destroy();
      wrapper = null;
    }
  });

  describe('when chartWidth is 0', () => {
    beforeEach(() => {
      wrapper = createComponent({
        query: mockProjectSecurityChartsWithData(),
        propsData: { hasVulnerabilities: true },
        chartWidth: 0,
      });
      return wrapper.vm.$nextTick();
    });

    it('should not display the line chart', () => {
      expect(findLineChart().exists()).toBe(false);
    });

    it('should display a loading icon instead', () => {
      expect(findLoadingIcon().exists()).toBe(true);
    });
  });

  describe('when there is history data', () => {
    beforeEach(() => {
      wrapper = createComponent({
        query: mockProjectSecurityChartsWithData(),
        propsData: { hasVulnerabilities: true },
      });
      return wrapper.vm.$nextTick();
    });

    it('should display the chart with data', async () => {
      expect(findLineChart().props('data')).toMatchSnapshot();
    });

    it('should not display the loading icon', () => {
      expect(findLoadingIcon().exists()).toBe(false);
    });
  });

  describe('when there is no history data', () => {
    beforeEach(() => {
      wrapper = createComponent({
        query: mockProjectSecurityChartsWithoutData(),
        propsData: { hasVulnerabilities: false },
      });
      return wrapper.vm.$nextTick();
    });

    it('should display the empty state', () => {
      expect(findEmptyState().props()).toEqual({ helpPath });
    });

    it('should not display the chart', () => {
      expect(findLineChart().exists()).toBe(false);
    });

    it('should not display the loading icon', () => {
      expect(findLoadingIcon().exists()).toBe(false);
    });
  });

  describe('when the query is loading', () => {
    beforeEach(() => {
      wrapper = createComponent({
        query: () => ({}),
        propsData: { hasVulnerabilities: true },
      });
    });

    it('should not display the empty state', () => {
      expect(findEmptyState().exists()).toBe(false);
    });

    it('should not display the chart', () => {
      expect(findLineChart().exists()).toBe(false);
    });

    it('should display the loading icon', () => {
      expect(findLoadingIcon().exists()).toBe(true);
    });
  });
});
