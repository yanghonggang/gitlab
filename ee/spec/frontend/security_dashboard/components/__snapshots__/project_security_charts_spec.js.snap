// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Security Charts component when there is history data should display the chart with data 1`] = `
Array [
  Object {
    "data": Array [
      Array [
        "2020-07-22",
        4,
      ],
      Array [
        "2020-07-23",
        2,
      ],
      Array [
        "2020-07-24",
        2,
      ],
      Array [
        "2020-07-25",
        2,
      ],
      Array [
        "2020-07-26",
        2,
      ],
      Array [
        "2020-07-27",
        2,
      ],
    ],
    "itemStyle": Object {
      "color": "#660e00",
    },
    "key": "critical",
    "lineStyle": Object {
      "color": "#660e00",
    },
    "name": "Critical",
  },
  Object {
    "data": Array [
      Array [
        "2020-07-22",
        3,
      ],
      Array [
        "2020-07-23",
        3,
      ],
      Array [
        "2020-07-24",
        3,
      ],
      Array [
        "2020-07-25",
        3,
      ],
      Array [
        "2020-07-26",
        3,
      ],
      Array [
        "2020-07-27",
        3,
      ],
    ],
    "itemStyle": Object {
      "color": "#ae1800",
    },
    "key": "high",
    "lineStyle": Object {
      "color": "#ae1800",
    },
    "name": "High",
  },
  Object {
    "data": Array [
      Array [
        "2020-07-22",
        2,
      ],
      Array [
        "2020-07-23",
        2,
      ],
      Array [
        "2020-07-24",
        2,
      ],
      Array [
        "2020-07-25",
        2,
      ],
      Array [
        "2020-07-26",
        2,
      ],
      Array [
        "2020-07-27",
        2,
      ],
    ],
    "itemStyle": Object {
      "color": "#9e5400",
    },
    "key": "medium",
    "lineStyle": Object {
      "color": "#9e5400",
    },
    "name": "Medium",
  },
  Object {
    "data": Array [
      Array [
        "2020-07-22",
        10,
      ],
      Array [
        "2020-07-23",
        10,
      ],
      Array [
        "2020-07-24",
        10,
      ],
      Array [
        "2020-07-25",
        10,
      ],
      Array [
        "2020-07-26",
        10,
      ],
      Array [
        "2020-07-27",
        10,
      ],
    ],
    "itemStyle": Object {
      "color": "#c17d10",
    },
    "key": "low",
    "lineStyle": Object {
      "color": "#c17d10",
    },
    "name": "Low",
  },
  Object {
    "data": Array [
      Array [
        "2020-07-22",
        1,
      ],
      Array [
        "2020-07-23",
        1,
      ],
      Array [
        "2020-07-24",
        1,
      ],
      Array [
        "2020-07-25",
        1,
      ],
      Array [
        "2020-07-26",
        1,
      ],
      Array [
        "2020-07-27",
        1,
      ],
    ],
    "itemStyle": Object {
      "color": "#868686",
    },
    "key": "unknown",
    "lineStyle": Object {
      "color": "#868686",
    },
    "name": "Unknown",
  },
  Object {
    "data": Array [
      Array [
        "2020-07-22",
        2,
      ],
      Array [
        "2020-07-23",
        2,
      ],
      Array [
        "2020-07-24",
        2,
      ],
      Array [
        "2020-07-25",
        2,
      ],
      Array [
        "2020-07-26",
        2,
      ],
      Array [
        "2020-07-27",
        2,
      ],
    ],
    "itemStyle": Object {
      "color": "#428fdc",
    },
    "key": "info",
    "lineStyle": Object {
      "color": "#428fdc",
    },
    "name": "Info",
  },
]
`;
